### ISSUE TITLE
Resource lookup errors when using shared libraries

### ISSUE THREAD

**Issue by bplevin36** [tensorflow/tensorflow#48058]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 20.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): bug occurs with both
- TensorFlow version (use command below): 2.3.2 CPU
- Python version: 3.10
- Bazel version (if compiling from source): 3.1.0
- GCC/Compiler version (if compiling from source): 9.3.0
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**Describe the current behavior**
Using the C API (`libtensorflow.so` and `libtensorflow_framework.so`) in conjunction with tensorflow-text (or any library loading custom ops) throws errors like:
`Invalid argument: Trying to access resource using the wrong type. Expected N10tensorflow6lookup15LookupInterfaceE got N10tensorflow6lookup15LookupInterfaceE`
This is caused by the same bug as issue #44209.  In `tensorflow/core/framework/type_index.h`, the address of a static variable is used as a hash, leading to a false mismatch when called from a shared library.  That poster suggested that the bug did not occur on Linux because of the GNU unique symbol extension.  In my case, it still happens even on Linux.

Looking at the shared libraries, I think I can see why:
```
$ nm -C /usr/local/lib/libtensorflow.so | grep LookupInterface.*hash_bit
0000000008e091ac b tensorflow::TypeIndex::Make<tensorflow::lookup::LookupInterface>()::hash_bit
                 ^ BSS symbol

$ nm -C /usr/local/lib/libtensorflow_framework.so | grep LookupInterface.*hash_bit
0000000001c656ec u tensorflow::TypeIndex::Make<tensorflow::lookup::LookupInterface>()::hash_bit
                 ^ UNIQUE symbol
```
In libtensorflow_framework.so, the hash_bit is a unique symbol, but in libtensorflow.so, it is not.  I wondered why this problem was not happening when running Tensorflow from Python, so I checked the Python shared library:
```
$ nm -C ./_pywrap_tensorflow_internal.so | grep LookupInterface.*hash_bit
000000001de2842c u tensorflow::TypeIndex::Make<tensorflow::lookup::LookupInterface>()::hash_bit
                 ^ UNIQUE symbol
```
Sure enough, unique symbol.  So the question is why does libtensorflow.so get this symbol in BSS when every other library has it as unique?  I don't know enough about Bazel to inspect how these libraries are being linked, so I'm hoping someone here can help.

**Describe the expected behavior**
This error should not be thrown.

**Comment by bplevin36** [comment]
I was able to hack my way around this by taking the fix in `tensorflow/core/framework/type_index.h` that was done for MacOS and removing the `#ifdef` guard (enabling it on all platforms), then rebuilding the library with `bazel build --config=opt --cxxopt=-D_GLIBCXX_USE_CXX11_ABI=0 tensorflow:libtensorflow.so`.  But this kind of hack really shouldn't be necessary.

**Comment by amahendrakar** [comment]
@bplevin36,
TensorFlow v2.3 is not compatible with Python 3.10. Please take a look at the [tested build configurations](https://www.tensorflow.org/install/source#cpu) for more information. 

Version | Python version | Compiler | Build tools
-- | -- | -- | --
tensorflow-2.4.0 | 3.6-3.8 | GCC 7.3.1 | Bazel 3.1.0
tensorflow-2.3.0 | 3.5-3.8 | GCC 7.3.1 | Bazel 3.1.0
tensorflow-2.2.0 | 3.5-3.8 | GCC 7.3.1 | Bazel 2.0.0

Could you please check if you are facing the same issue with Python 3.8 as well? Thanks!

**Comment by bplevin36** [comment]
@amahendrakar Thanks for the quick response.  I included my Python version just in case Python was somehow being used by the build process, but my actual program is written in Rust and includes no Python.  It only uses the shared libraries.

**Comment by bplevin36** [comment]
@amahendrakar  Just in case, I reran the entire process using Python 3.8, but the issue remains.

**Comment by maxhgerlach** [comment]
@bplevin36 @jvishnuvardhan, I've run into this issue with TensorFlow 2.4. Is this fixed in later versions?

**Comment by maxhgerlach** [comment]
This might be fixed in TF 2.6 via this PR https://github.com/tensorflow/tensorflow/pull/47072

**Comment by Venkat6871** [comment]
Hi @bplevin36 ,
Could you please check with latest versions, if you are facing the same issue? Please take a look at the [installation](https://www.tensorflow.org/install/source#cpu) process for more information.  

Thank you!
